google/cloud/internal/big_endian.h

```diff
@@ -15,164 +15,67 @@
-                                 (((num_value)&0x00ff000000000000ull) >> 40) |
-                                 (((num_value)&0x0000ff0000000000ull) >> 24) |
-                                 (((num_value)&0x000000ff00000000ull) >> 8) |
-                                 (((num_value)&0x00000000ff000000ull) << 8) |
-                                 (((num_value)&0x0000000000ff0000ull) << 24) |
-                                 (((num_value)&0x000000000000ff00ull) << 40) |
-                                 (((num_value)&0x00000000000000ffull) << 56));
-
-    return return_value;
-#endif
-  }
-
-  static inline std::uint16_t byte_swap(std::uint16_t value) {
-#ifdef _MSC_VER
-    return _byteswap_ushort(value);
-#elif defined(__APPLE__)
-    return OSSwapInt16(value);
-#elif defined(__GNUC__) || defined(__clang__)
-    return __builtin_bswap16(value);
-#else
-    std::uint16_t return_value =
-        ((((num_value)&0xff00u) >> 8) | (((num_value)&0x00ffu) << 8));
 
-    return return_value;
-#endif
-  }
-
-  static inline std::uint32_t byte_swap(std::uint32_t value) {
-#ifdef _MSC_VER
-    return _byteswap_ulong(value);
-#elif defined(__APPLE__)
-    return OSSwapInt32(value);
-#elif defined(__GNUC__) || defined(__clang__)
-    return __builtin_bswap32(value);
-#else
-    std::uint32_t return_value =
-        ((((num_value)&0xff000000ul) >> 24) |
-         (((num_value)&0x00ff0000ul) >> 8) | (((num_value)&0x0000ff00ul) << 8) |
-         (((num_value)&0x000000fful) << 24));
-
-    return return_value;
-#endif
-  }
-
-  static inline std::uint64_t byte_swap(std::uint64_t value) {
-#ifdef _MSC_VER
-    return _byteswap_uint64(value);
-#elif defined(__APPLE__)
-    return OSSwapInt64(value);
-#elif defined(__GNUC__) || defined(__clang__)
-    return __builtin_bswap64(value);
-#else
-    std::uint64_t return_value = ((((num_value)&0xff00000000000000ull) >> 56) |
-                                  (((num_value)&0x00ff000000000000ull) >> 40) |
-                                  (((num_value)&0x0000ff0000000000ull) >> 24) |
-                                  (((num_value)&0x000000ff00000000ull) >> 8) |
-                                  (((num_value)&0x00000000ff000000ull) << 8) |
-                                  (((num_value)&0x0000000000ff0000ull) << 24) |
-                                  (((num_value)&0x000000000000ff00ull) << 40) |
-                                  (((num_value)&0x00000000000000ffull) << 56));
-
-    return return_value;
-#endif
+// Encodes an 8-bit, 16-bit, 32-bit, or 64-bit signed or unsigned value as a
+// big-endian sequence of bytes. The returned string has a size matching
+// `sizeof(T)`. Example:
+//
+//   std::string s = EncodeBigEndian(std::int32_t{255});
+//   assert(s == std::string("\0\0\0\xFF", 4));
+//
+template <typename T,
+          typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+inline std::string EncodeBigEndian(T value) {
+  static_assert(std::numeric_limits<unsigned char>::digits == 8,
```

Comment:
s/unsigned char/std::string::value_type/?

Reply:
I tried that originally, the problem is that sometimes char is unsigned, so its digits will be `CHAR_BIT - 1`. Really, we want to check that char has 8 bits. I could just check `CHAR_BIT` directly. Preference?

Reply:
Would that work though?  On a `CHAR_BIT==16` platform how does gRPC handle a string (or really a `bytes` field)?  I would expect that each octet received from the wire is stored into a different `char`, as opposed to nicely packing all the octets in half as many chars?  If we do not know how (or *if*) gRPC works on platforms with `CHAR_BIT==16` maybe we are better off not guessing?

Reply:
Sure ... if the string is really a sequence of octets I'd also be happy to remove the assert and use a constant 8.  What goes over the wire, though, should be somewhat independent of how values are represented in C++ on a machine.  But I'll try to learn more about what these big-endian things are before commenting further.

Reply:
I think we are agreeing, maybe?  I do not know how gRPC+protobuf would chose to represent the underlying field:

https://github.com/googleapis/googleapis/blob/668b4f35a225145f1231c848be879ee950d4d203/google/bigtable/v2/data.proto#L82

on a platform where `CHAR_BIT==16`.  The [documentation](https://developers.google.com/protocol-buffers/docs/proto3#scalar) says that protobuf `bytes` are mapped to `std::string` in C++.  But the documentation is silent about the impact of `CHART_BIT` on this mapping.

They might use only 8 bits of each `char`, or they might pack 2 octets on each `char` (in what order?), or (possibly) protobuf and or gRPC do not support such platforms at all.

Without knowing that I think a static assert is the nicest thing we can do for our users.

Reply:
> Without knowing that I think a static assert is the nicest thing we can do for our users.

Ack.  I'm happy to wait for someone to complain that the assert fails for them, knowing that this will never actually happen.  I just dislike asserts that the code shouldn't really depend on.